index.bs

```diff
@@ -658,20 +658,20 @@ When the <dfn method for="XRFrame">getViewerPose(|referenceSpace|)</dfn> method
   1. Let |session| be the {{XRFrame}}'s {{XRFrame/session}} object.
   1. If |referenceSpace|'s [=XRSpace/session=] does not equal |session|, throw an {{InvalidStateError}} and abort these steps.
   1. If the [=viewer=]'s pose cannot be determined relative to |referenceSpace|, return <code>null</code>
-  1. Return a new {{XRViewerPose}} describing the [=viewer=]'s pose relative to the origin of |referenceSpace| at the timestamp of the {{XRFrame}}.
+  1. Return a new {{XRViewerPose}} with an {{XRRigidTransform}} from the [=viewer=] space to |referenceSpace| at the timestamp of the {{XRFrame}}. The transform's {{XRRigidTransform/position}} is the location of the [=viewer=] in reference space coordinates.
 
 </div>
 
 <div class="algorithm unstable" data-algorithm="get-pose">
 
-When the <dfn method for="XRFrame">getPose(|space|, |relativeTo|)</dfn> method is invoked, the user agent MUST run the following steps:
+When the <dfn method for="XRFrame">getPose(|sourceSpace|, |destinationSpace|)</dfn> method is invoked, the user agent MUST run the following steps:
 
   1. If the {{XRFrame}}'s [=active=] boolean is <code>false</code>, throw a {{InvalidStateError}} and abort these steps.
   1. Let |session| be the {{XRFrame}}'s {{XRFrame/session}} object.
-  1. If |space|'s [=XRSpace/session=] does not equal |session|, throw an {{InvalidStateError}} and abort these steps.
-  1. If |relativeTo|'s [=XRSpace/session=] does not equal |session|, throw an {{InvalidStateError}} and abort these steps.
-  1. If |space|'s pose cannot be determined relative to |relativeTo|, return <code>null</code>
-  1. Return a new {{XRPose}} describing |space|'s pose relative to the origin of |relativeTo|.
+  1. If |destinationSpace|'s [=XRSpace/session=] does not equal |session|, throw an {{InvalidStateError}} and abort these steps.
+  1. If |sourceSpace|'s [=XRSpace/session=] does not equal |session|, throw an {{InvalidStateError}} and abort these steps.
+  1. If |destinationSpace|'s pose cannot be determined relative to |sourceSpace|, return <code>null</code>
+  1. Return a new {{XRPose}} containing the {{XRRigidTransform}} from |sourceSpace| to |destinationSpace|. The transform's {{XRRigidTransform/position}} is the location of the |sourceSpace| origin in |destinationSpace| coordinates.
```

Comment:
Does anyone have good naming suggestions here? "sourceSpace" and "destinationSpace" seemed like the best fit for describing the transform, but it's a bit clunky when talking about getPose.

Do any of these variants look better? I've added snippets how the naming would affect this paragraph and the XRRigidTransform definition below.

`getPose(XRSpace sourceSpace, XRSpace destinationSpace);` =>
> Return a new {{XRPose}} containing the {{XRRigidTransform}} from |sourceSpace| to |destinationSpace|. The transform's {{XRRigidTransform/position}} is the location of the |sourceSpace| origin in |destinationSpace| coordinates.
>
> An {{XRRigidTransform}} is a transform from a source {{XRSpace}} to a destination {{XRSpace}} described by ...

`getPose(XRSpace space, XRSpace relativeTo);` =>
> Return a new {{XRPose}} containing the {{XRRigidTransform}} from |space| to |relativeTo|. The transform's {{XRRigidTransform/position}} is the location of the |space| origin in |relativeTo| coordinates.
>
> An {{XRRigidTransform}} is a transform from an {{XRSpace}} to a relativeTo {{XRSpace}} described by ...

`getPose(XRSpace targetSpace, XRSpace baseSpace);` =>
> Return a new {{XRPose}} containing the {{XRRigidTransform}} from |targetSpace| to |baseSpace|. The transform's {{XRRigidTransform/position}} is the location of the |targetSpace| origin in |baseSpace| coordinates.
>
> An {{XRRigidTransform}} is a transform from a target {{XRSpace}} to a base {{XRSpace}} described by ...

`getPose(XRSpace transformedSpace, XRSpace baseSpace);` =>
> Return a new {{XRPose}} containing the {{XRRigidTransform}} from |transformedSpace| to |baseSpace|. The transform's {{XRRigidTransform/position}} is the location of the |transformedSpace| origin in |baseSpace| coordinates.
>
> An {{XRRigidTransform}} is a transform from a transformed {{XRSpace}} to a base {{XRSpace}} described by ... _(This one sounds backwards)_

Reply:
@NellWaliczek please see [^^^](https://github.com/immersive-web/webxr/pull/569/files#r277060085), trying to get the naming sorted out in relation to your request from https://github.com/immersive-web/webxr/pull/589#discussion_r277049962

Reply:
I hope to somewhat redundantly clarify `getPose` with some matrix math so that the confusion doesn't get too bad. I kinda like `source` `destination`.

Reply:
Of the naming options above, I prefer `targetSpace`/`baseSpace`.  Specifically, I worry that `sourceSpace`/`destinationSpace` would present a reversed mental model to the way in which `getPose` will be increasingly used in practice.

As we start to establish plane anchors, mesh anchors, face anchors, etc., there will be an increasing number of static and dynamic entities in WebXR that are tracked over time by `XRSpace` instances.  To place these entities within their scene each frame, developers will use `xrFrame.getPose(interestingEntitySpace, myReferenceSpace)` and expect to get back the position and orientation of that entity within their engine's world coordinates (which are underwritten by `myReferenceSpace`), so they can update that entity's scene object transform in their engine.  The mental model for `getPose` in particular will start to become "what's the position and orientation of this single entity in my reference space", moreso than "get me a model transform I can use to generally transform coordinates from this entity space into my reference space".

However, the above is ultimately just me arguing that we should preference naming that's intuitive for the `position`/`orientation` attributes on `XRRigidTransform` rather than the `matrix` attribute.  In the end, @klausw had a key observation in https://github.com/immersive-web/webxr/issues/580#issuecomment-479595386 that applies here:

> I'm starting to think that a core part of the confusion is that a tracked object's XRPose has-a XRRigidTransform, and the transform's position component corresponds to the tracked object's coordinates in reference space as expected, but a XRRigidTransform and its matrix are a transform from object space to reference space. That's internally consistent, but it's confusing when you supply a transform by itself as an argument (as in the constructor arg here) because it's easily interpreted as saying "please apply this transform" which would be backwards. In a way, OpenXR's approach of directly putting position + orientation on the pose objects helps avoid this ambiguity.

Ultimately, **any naming pattern and/or conceptual framework we come up with that makes `position`/`orientation` make more sense will make `matrix` make less sense, and vice versa**.

In OpenXR, we've managed to sidestep entire classes of confusion around the order or direction in which matrices are applied, whether they're pre-multiplied or post-multiplied, whether they are row-major/column-major, etc. by **banishing matrices from the API altogether**.  Rather than expressing transform matrices "from" or "to" various spaces, all poses are expressed only as a `position` vector and `orientation` quaternion "in" a given space.

Even for view matrices, as discussed in #447, we found in OpenXR that any non-trivial engine ends up positioning a camera object within its scene in world coordinates anyway, and so even there it ends up being more convenient for developers to reason about the position and orientation of their view rather than getting a view matrix directly.  You can see this pattern in OpenXR in [xrLocateViews](https://www.khronos.org/registry/OpenXR/specs/0.90/man/html/#_xrlocateviews_3), which returns a set of [XrView](https://www.khronos.org/registry/OpenXR/specs/0.90/man/html/#_xrview_3) structs, each containing the position and orientation of a given view.

Specifically in the case of OpenXR reference spaces with a non-identity pose applied:
* Each reference space type is specified in terms of how its "natural" origin behaves.
* If you apply a non-identity pose when creating a reference space, that's specified as placing the origin of your reference space at that position and orientation "in" the "natural" reference space.

By defining the origin of any space as either a well-specified location in the physical world or a position and orientation within some other such space, all space origins have a clear binding to a place in the physical world.  It then falls out without ambiguity how you "get" the "pose" of any one space "in" another.

If we ourselves continue to struggle with these naming and direction issues each week, despite being embedded in the details of the spec, perhaps it's the lesser evil to just remove `matrix` from `XRRigidTransform`?  This will forcibly snap the developer's mental model to an unambiguous position and orientation approach throughout the API, and then we can design our naming to be consistent with that model.

Reply:
I think the problem here isn't `matrix`, the problem is the use of the word `transform` itself.

I point this out here: https://github.com/immersive-web/editor-collab/pull/36#discussion_r278361579 , there are two ways one can look at a transform. _Just_ talking about position/orientation as a "pose" is clear, however when you talk of the "transform" of a position/orientation that can mean opposite things.

If we can carefully clarify our use of the word "transform", talking about matrices isn't hard.

I feel like we should definitely try and keep `matrix` around for developer convenience.